docs/source/expr-system-internals.rst

```diff
@@ -0,0 +1,206 @@
+Query planning with Expression system
+=====================================
+
+.. note::
+
+    This document is intended for Dask developers and contributors. It is not
+    intended for end-users.
+    For a high level user guide, see :doc:`/dataframe-optimizer`.
+
+.. currentmodule:: dask.dataframe
+
+The expression system was originally developed for Dask DataFrames, as implemented in the `dask-expr <https://github.com/dask/dask-expr>`_ project. Early prototypes experimented with `matchpy <https://github.com/HPAC/matchpy>`_, but it was soon replaced with a custom-built system offering simpler and more transparent optimization steps.
```

Comment:
I recommend dropping matchpy (not sure that the history is relevant)

Maybe also point out the larger objective beyond just dataframes?

Reply:
Maybe also good to mention motivations here (high level optimization, more compact representation to move around, etc..)

Reply:
I don't intend to motivate why this is here. If anything, this document is supposed to explain a couple of gotchas around the system. If one is looking for motivation or high level introduction, we have https://docs.dask.org/en/stable/dataframe-optimizer.html

Reply:
I think that motivating the expression system as a whole is still a good idea.  The linked doc mostly says "we do optimizations and here's what they are".  I think something higher level would be useful.  Not critical though.